Ser for msp podcast

Subohao opened the pull request:
## What?

Modify the template from slu1 and spk to ser downstream tasks, specifically for primary categorical emotions in msp-podcast, excluding X and O.

## Why?

For ESPnet BootCamp and integrating SER downstream tasks into ESPnet.

## See also

<!-- Write additional information if necessary (e.g., referecne, related PRs or Issues). -->

Commit:
adding templates for ser tasks

Commit:
adding examples for running msp_podcast on the primary speech emotion recongition task excluding O and X emotions

Subohao committed:
adding ser training and inference bin

On branch SER-for-msp-podcast
Changes to be committed:
	new file:   espnet2/bin/ser_inference.py
	new file:   espnet2/bin/ser_train.py

Subohao committed:
adding ser to espnet2

Subohao committed:
adding ser to tasks

Subohao committed:
adding SER preprocessor

Subohao committed:
remove some redundant parameters in SER

Subohao committed:
remove redundant parameters in SER

Subohao committed:
remove redundant parameters in SER

Subohao committed:
remove redundant parameters in SER

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

sw005320 commented:
Can you fix the CI error in https://github.com/espnet/espnet/actions/runs/14582314636/job/40901457851?pr=6096#step:8:285 ?

sw005320 commented:
I thought that you would implement it in the `slu` task.
Either is fine, but can you explain why you created a new task?

Subohao commented:
> I thought that you would implement it in the `slu` task. Either is fine, but can you explain why you created a new task?

Based on my understanding of the SLU recipe, it's tied to the ASR task as well. In this SER recipe, I certainly simplify it into a simple emotion classification task by only considering the speech input and emotions as output, without tackling much on the speaker and text part of MSP-PODCAST. Also, in SER, we would usually have a pooling and a projection layer at the end of the model, which does not seem to be in the SLU recipe. Therefore, I created a new task to do it.

Subohao committed:
remove unused variable

Subohao committed:
replace unused variable with _

sw005320 commented:
@ftshijt, can you review this PR since it is a new task.

ftshijt committed:
Merge branch 'master' into SER-for-msp-podcast

ftshijt commented on egs2/TEMPLATE/ser1/ser.sh:
```diff
@@ -0,0 +1,1677 @@
+                    --non_linguistic_symbols "${nlsyms_txt}" \
+                    --cleaner "${cleaner}" \
+                    --g2p "${g2p}" \
+                    --emotions "A_S_H_U_F_D_C_N_O_X" \
```
Is emotion fixed only to these categories?

Subohao replied:
Since now it's temporary for MSP_Podcast, it's only used for these emotion categories.

ftshijt commented on egs2/TEMPLATE/ser1/ser.sh:
```diff
@@ -0,0 +1,1677 @@
+                    --collect_stats true \
+                    --use_preprocessor true \
+                    --bpemodel "${bpemodel}" \
+                    --transcript_token_list "${transcript_token_list}" \
+                    --token_type "${token_type}" \
+                    --token_list "${token_list}" \
```
what is the difference between transcript token list and token list?

ftshijt commented on egs2/TEMPLATE/ser1/ser.sh:
```diff
@@ -0,0 +1,1677 @@
+
+
+if ! "${skip_train}"; then
+    if "${use_lm}"; then
+        if [ ${stage} -le 6 ] && [ ${stop_stage} -ge 6 ]; then
+            log "Stage 6: LM collect stats: train_set=${data_feats}/lm_train.txt, dev_set=${lm_dev_text}"
```
It seems that we do not need to have LM stages?

ftshijt commented on egs2/TEMPLATE/ser1/ser.sh:
```diff
@@ -0,0 +1,1677 @@
+                    --output_dir "${_logdir}/stats.JOB" \
+                    ${_opts} ${ser_args} || { cat "${_logdir}"/stats.1.log; exit 1; }
+        else
+                    # --bpemodel "${bpemodel}" \
+                    # --token_type "${token_type}" \
+                    # --token_list "${token_list}" \
+                    # --non_linguistic_symbols "${nlsyms_txt}" \
+                    # --cleaner "${cleaner}" \
+                    # --g2p "${g2p}" \
```
Please clean up the commented lines (also for following parts)

ftshijt commented on egs2/msp_podcast/ser1/cmd.sh:
```diff
@@ -0,0 +1,110 @@
+
+
+# Select the backend used by run.sh from "local", "stdout", "sge", "slurm", or "ssh"
+cmd_backend='slurm' #cmd_backend='local'
```
You would need to change it to local to pass the ci

ftshijt commented on egs2/msp_podcast/ser1/local/score_ser.py:
```diff
@@ -0,0 +1,52 @@
+import argparse
```
If it is general, you may consider putting it into pyscripts

Subohao replied:
Moved

ftshijt commented on egs2/msp_podcast/ser1/run.sh:
```diff
@@ -0,0 +1,52 @@
+#   emotional labels: ang (anger), hap (happiness), exc (excitement), sad (sadness),
+#   fru (frustration), fea (fear), sur (surprise), neu (neutral), and xxx (other)
+#--convert_to_sentiment
+#   Convert emotion to sentiment (Positive, Negative and Neutral)
+#   mapping from emotion to sentiment is as follows:
+#   Positive: hap, exc, sur
+#   Negative: ang, sad, fru, fea
+#   Neutral: neu
```
Given the different definition for each dataset, I guess you might need to have an option to control the categories of emotion?

ftshijt commented on egs2/msp_podcast/ser1/ser.sh:
```diff
@@ -0,0 +1,1677 @@
+#!/usr/bin/env bash
```
Instead of copying the ser, please use symlink

Subohao replied:
Done

ftshijt commented on espnet2/bin/ser_inference.py:
```diff
@@ -0,0 +1,306 @@
+        task = SERTask
+
+        # 1. Build SER model
+        scorers = {}
```
Is beamsize necessary for the application (also for scorers)

ftshijt commented on espnet2/ser/pooling/abs_pooling.py:
```diff
@@ -0,0 +1,13 @@
+from abc import ABC, abstractmethod
```
It is okay to reuse existing modules (if no significant changes require for the design)

Subohao replied:
Just create a symlink to the spk pooling function insteadly.

ftshijt reviewed (commented):
Many thanks for the great contribution. I left some comments as follows (mostly related to redundant components) -> if we choose to expand to another task from existing ones, one would be especially beneficial if we can make some reduction to the heavy process~

sw005320 committed:
Merge branch 'master' into SER-for-msp-podcast

Subohao committed:
Merge branch 'espnet:master' into SER-for-msp-podcast

Subohao committed:
update the description for README.md

Subohao committed:
change slurm to local for passing ci

Subohao committed:
using symlink to TEMPLATE

Subohao committed:
comment to test

Subohao committed:
remove beamsize arguments

Subohao committed:
reuse the pooling code from spk task, generate a symlink

Subohao committed:
move score_ser to pyscript and change stage 13 command accordingly

Subohao committed:
move score_ser to pyscript

Subohao committed:
remove LM-related part in the script

sw005320 commented:
Please fix the errors
https://github.com/espnet/espnet/actions/runs/15715129680/job/44282859093?pr=6096

Subohao committed:
Merge branch 'espnet:master' into SER-for-msp-podcast

Subohao committed:
remove some unused packages in ser.py

Subohao committed:
remove unsed packages in cross entropy loss

Subohao committed:
remove unsed packages in ser espnet model

Subohao committed:
remove unsed packages in ser_inference

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Subohao committed:
Merge branch 'espnet:master' into SER-for-msp-podcast

Subohao committed:
cp template slrum.conf to egs2/msp_podcast

Subohao committed:
add unit test code for SER, tested locally

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Subohao committed:
Merge branch 'espnet:master' into SER-for-msp-podcast

sw005320 commented:
The test took too long.
https://github.com/espnet/espnet/actions/runs/15882286538/job/44785536230?pr=6096#step:9:3174

Please do either of them
1. Reduce the time by using a smaller model or shorter file (recommended)
2. Increase the test time.

Subohao committed:
add a dummpy data part for integration unit test

Subohao committed:
create a folder for integration unit-test

Subohao committed:
add ser in the integration shell

Subohao committed:
change to a smaller model and shorter file

Subohao committed:
add an extra argument for running dummy data generation to pass unit-test

Subohao committed:
change to smaller model and shorter file as well

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Subohao committed:
add quote to split sets

Subohao committed:
Merge branch 'SER-for-msp-podcast' of https://github.com/Subohao/espnet into SER-for-msp-podcast

Subohao committed:
update the emotion label to fit smaller model

sw005320 committed:
make CI happy

sw005320 commented on egs2/mini_an4/ser1/conf/tuning/train_asr_wavlm.yaml:
I'm concerned that this config would be too much as an integration test.
If it were a problem in the future, I'd request that you remove the `wavlm` part

sw005320 commented on espnet2/ser/loss/cross_entropy_loss.py:
```diff
@@ -0,0 +1,37 @@
+
+
+class Xnt(AbsLoss):
+    """Additive angular margin softmax.
+
+    Paper: Deng, Jiankang, et al. "Arcface: Additive angular margin loss for
+    deep face recognition." Proceedings of the IEEE/CVF conference on computer
+    vision and pattern recognition. 2019.
```
Is it an angular margin softmax?
It looks like a normal softmax function.

Subohao replied:
Just remove the incorrect description in the function.

sw005320 commented on espnet2/ser/pooling/mean_pooling.py:
Did you care about the padded region?

sw005320 commented on espnet2/tasks/ser.py:
```diff
@@ -0,0 +1,396 @@
+from espnet2.asr.specaug.abs_specaug import AbsSpecAug
+from espnet2.asr.specaug.specaug import SpecAug
+from espnet2.ser.espnet_model import ESPnetSERModel
+from espnet2.ser.loss.cross_entropy_loss import Xnt
+from espnet2.ser.pooling.abs_pooling import AbsPooling
+from espnet2.ser.pooling.mean_pooling import MeanPooling
+from espnet2.ser.projector.abs_projector import AbsProjector
+from espnet2.ser.projector.linear_projector import LinearProjector
```
Many functions are similar to those in SID.
Please consider reusing them.

Subohao replied:
I create a symbolic link to the SID if there exists same functions.

sw005320 commented on espnet2/train/preprocessor.py:
```diff
@@ -638,6 +638,64 @@ def _text_process(
         return data
 
 
+class SERPreprocessor(CommonPreprocessor):
```
Do you need a SER-specific preprocessor?
If so, can you explain it and add it to the comment?

Subohao replied:
Create a commit to include the reason for SER-specific preprocessor.

sw005320 committed:
Merge branch 'master' into SER-for-msp-podcast

sw005320 commented:
@Subohao, please work on this PR

Subohao committed:
Merge branch 'espnet:master' into SER-for-msp-podcast

Subohao committed:
reomve uncorrect description of cross-entropy loss

Subohao committed:
taking care of padded region, which is different from SID in the dimension of mean axis

Subohao committed:
resuing abs_projector from SID by creating symbolic link

Subohao committed:
redo

Subohao committed:
Since the SER need to encode and map the emotion, I create a new preprocessor to take care the of emo label in the task

Subohao committed:
add comment in the preprocessing to indicate the necessity of SER-specific preprocessor

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Subohao committed:
fix comment too ling issue in unit test

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Subohao committed:
to do ci again

Subohao committed:
fix git issue

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Subohao committed:
fix bugs

Subohao committed:
Merge branch 'espnet:master' into SER-for-msp-podcast

Subohao committed:
fix bugs in ci and add an extra file to show ser result to final README

Subohao committed:
new file to show ser result in README

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Subohao committed:
do ci again

Subohao committed:
ci

Subohao committed:
fix the stage

Subohao committed:
Merge branch 'espnet:master' into SER-for-msp-podcast

Subohao committed:
Merge branch 'espnet:master' into SER-for-msp-podcast

Subohao committed:
modifications according to code review comments

Subohao committed:
rerun_CI

Subohao committed:
rerun CI

Subohao committed:
Merge branch 'espnet:master' into SER-for-msp-podcast

Subohao committed:
Use wavlm_base_plus for CI testing instead of wavlm_large

Subohao committed:
solve torch backend issue

Subohao committed:
Add CNN frontend option for CI testing

  - Add CNNFrontend to SER task frontend choices
  - Create lightweight CNN-based config for CI testing
  - Avoid s3prl dependency and model downloading issues
  - Use simple 3-layer CNN with 256-dim output for mini_an4

Subohao committed:
avoid ci test error

Subohao committed:
Add comprehensive test coverage for SER (Speech Emotion Recognition) module

This commit adds extensive test coverage to address the Codecov patch failure
by testing previously uncovered code paths in the SER implementation.

Added tests for:
- SER model methods: get_prediction, collect_feats, extract_feats, encode
- SER model with optional components (without preencoder/frontend)
- SER inference: Speech2Emotion class and inference function
- SER training entry points: get_parser and main
- SER task: argument parsing and configuration

These tests ensure that the new SER functionality has adequate code coverage
to meet Codecov requirements.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Subohao committed:
Fix flake8 error: remove unused variable in test_ser_without_frontend

Subohao committed:
Fix test: correct ModelDownloader patch path in ser_inference test

Subohao committed:
Merge branch 'espnet:master' into SER-for-msp-podcast

Subohao committed:
fix inference error

Subohao committed:
re-run again, since local is ok

Subohao committed:
update the configuration

Subohao committed:
fix inference error

Subohao commented:
@sw005320 Would you mind checking the PR  and merge it when you are available? After modifying the errors and reflecting the comments, the PR finally passed all the CI test. Thank you!

Subohao committed:
reflect gemini's comments

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Subohao committed:
update due to gemini's comments adjustment
